compiler/noirc_evaluator/src/ssa/opt/remove_unreachable.rs

```diff
@@ -33,66 +34,35 @@ use crate::ssa::{
 impl Ssa {
     /// See [`remove_unreachable`][self] module for more information.
     pub(crate) fn remove_unreachable_functions(mut self) -> Self {
-        let mut reachable_functions = HashSet::default();
-
-        // Go through all the functions, and if we have an entry point, extend the set of all
-        // functions which are reachable.
-        for (id, function) in self.functions.iter() {
+        // Identify entry points
+        let entry_points = self.functions.iter().filter_map(|(&id, func)| {
             // Not using `Ssa::is_entry_point` because it could leave Brillig functions that nobody calls in the SSA,
             // because it considers every Brillig function as an entry point.
-            let is_entry_point = function.id() == self.main_id
-                || function.runtime().is_acir() && function.runtime().is_entry_point();
-
-            if is_entry_point {
-                collect_reachable_functions(&self, *id, &mut reachable_functions);
-            }
-        }
+            let is_entry_point =
+                id == self.main_id || func.runtime().is_acir() && func.runtime().is_entry_point();
+            is_entry_point.then_some(id)
+        });
+
+        // Build call graph dependencies using this passes definition of reachability.
+        let dependencies = self
+            .functions
+            .iter()
+            .map(|(&id, func)| {
+                let used = used_functions(func);
+                (id, used)
+            })
+            .collect();
```

Comment:
I think this is the result:

```suggestion
        let dependencies =
            self.functions.iter().map(|(&id, func)| (id, used_functions(func))).collect();
```